File: templates/ts/protocols/mavlink_vehicle_client.ts
import type { MavlinkDialect } from '../mavlink_dialect';
import { HeartbeatTemplates } from './heartbeat_protocol';
import type { MavlinkLink } from './mavlink_link';
import { MavlinkSession } from './mavlink_session';
import { MavlinkNode, HeartbeatMonitor, HeartbeatPublisher } from './heartbeat_protocol';
import { CommandProtocol } from './command_protocol';
import { MissionProtocol } from './mission_protocol';
import { ParameterProtocol } from './parameter_protocol';

/** Protocol clients bound to a single remote MAVLink vehicle. */
export class MavlinkVehicleClient {
  readonly session: MavlinkSession;
  readonly vehicle: MavlinkNode;
  readonly parameters: ParameterProtocol;
  readonly mission: MissionProtocol;
  readonly command: CommandProtocol;

  constructor(options: {
    session: MavlinkSession;
    vehicle: MavlinkNode;
    parameterRequestTimeoutMs?: number;
    parameterIdleTimeoutMs?: number;
    missionItemTimeoutMs?: number;
    missionOperationTimeoutMs?: number;
    commandTimeoutMs?: number;
  }) {
    this.session = options.session;
    this.vehicle = options.vehicle;
    this.parameters = new ParameterProtocol({
      session: options.session,
      targetSystem: options.vehicle.systemId,
      targetComponent: options.vehicle.componentId,
      requestTimeoutMs: options.parameterRequestTimeoutMs ?? 10_000,
      idleTimeoutMs: options.parameterIdleTimeoutMs ?? 2000,
    });
    this.mission = new MissionProtocol({
      session: options.session,
      targetSystem: options.vehicle.systemId,
      targetComponent: options.vehicle.componentId,
      itemTimeoutMs: options.missionItemTimeoutMs ?? 10_000,
      operationTimeoutMs: options.missionOperationTimeoutMs ?? 30_000,
    });
    this.command = new CommandProtocol({
      session: options.session,
      targetSystem: options.vehicle.systemId,
      targetComponent: options.vehicle.componentId,
      defaultTimeoutMs: options.commandTimeoutMs ?? 10_000,
    });
  }

  get targetSystem(): number {
    return this.vehicle.systemId;
  }

  get targetComponent(): number {
    return this.vehicle.componentId;
  }
}

/** Ground control station bootstrap: session, heartbeat publisher, and monitor. */
export class MavlinkGcs {
  readonly session: MavlinkSession;
  readonly heartbeatPublisher: HeartbeatPublisher;
  readonly heartbeatMonitor: HeartbeatMonitor;

  constructor(options: {
    session: MavlinkSession;
    heartbeatPublisher: HeartbeatPublisher;
    heartbeatMonitor: HeartbeatMonitor;
  }) {
    this.session = options.session;
    this.heartbeatPublisher = options.heartbeatPublisher;
    this.heartbeatMonitor = options.heartbeatMonitor;
  }

  /** Start heartbeat publish/monitor loops. */
  start(): void {
    this.heartbeatMonitor.start();
    this.heartbeatPublisher.start();
  }

  /** Stop heartbeat publish/monitor loops. */
  async stopHeartbeats(): Promise<void> {
    this.heartbeatPublisher.stop();
    await this.heartbeatMonitor.stop();
  }

  /** Wait for the first vehicle and return a [MavlinkVehicleClient] for it. */
  async waitForVehicle(options: {
    excludeSystemIds?: ReadonlySet<number>;
    timeoutMs?: number;
  } = {}): Promise<MavlinkVehicleClient> {
    const node = await this.heartbeatMonitor.waitForVehicle(options);
    return new MavlinkVehicleClient({ session: this.session, vehicle: node });
  }

  /** Build a [MavlinkVehicleClient] for a known [vehicle] node. */
  vehicleClient(vehicle: MavlinkNode): MavlinkVehicleClient {
    return new MavlinkVehicleClient({ session: this.session, vehicle });
  }

  /** Factory for a typical GCS setup over [link]. */
  static connect(options: {
    dialect: MavlinkDialect;
    link: MavlinkLink;
    systemId?: number;
    componentId?: number;
    heartbeatIntervalMs?: number;
    heartbeatTimeoutMs?: number;
  }): MavlinkGcs {
    const session = new MavlinkSession({
      dialect: options.dialect,
      link: options.link,
      systemId: options.systemId ?? 255,
      componentId: options.componentId ?? 190,
    });

    return new MavlinkGcs({
      session,
      heartbeatPublisher: new HeartbeatPublisher({
        session,
        heartbeat: HeartbeatTemplates.gcs(options.dialect.version),
        intervalMs: options.heartbeatIntervalMs ?? 1000,
      }),
      heartbeatMonitor: new HeartbeatMonitor({
        session,
        timeoutMs: options.heartbeatTimeoutMs ?? 3000,
      }),
    });
  }

  async close(): Promise<void> {
    await this.stopHeartbeats();
    await this.session.close();
  }
}

File: templates/dart/protocols/mavlink_vehicle_client.dart
import '../mavlink_dialect.dart';
import 'command_protocol.dart';
import 'heartbeat_protocol.dart';
import 'mavlink_link.dart';
import 'mavlink_session.dart';
import 'mission_protocol.dart';
import 'parameter_protocol.dart';

/// Protocol clients bound to a single remote MAVLink vehicle.
///
/// Created after a vehicle is discovered (e.g. via [HeartbeatMonitor.waitForVehicle]).
/// Shares one [MavlinkSession] and consistent target addressing for all services.
class MavlinkVehicleClient {
  MavlinkVehicleClient({
    required this.session,
    required this.vehicle,
    Duration parameterRequestTimeout = const Duration(seconds: 10),
    Duration parameterIdleTimeout = const Duration(seconds: 2),
    Duration missionItemTimeout = const Duration(seconds: 10),
    Duration missionOperationTimeout = const Duration(seconds: 30),
    Duration commandTimeout = const Duration(seconds: 10),
  })  : parameters = ParameterProtocol(
          session: session,
          targetSystem: vehicle.systemId,
          targetComponent: vehicle.componentId,
          requestTimeout: parameterRequestTimeout,
          idleTimeout: parameterIdleTimeout,
        ),
        mission = MissionProtocol(
          session: session,
          targetSystem: vehicle.systemId,
          targetComponent: vehicle.componentId,
          itemTimeout: missionItemTimeout,
          operationTimeout: missionOperationTimeout,
        ),
        command = CommandProtocol(
          session: session,
          targetSystem: vehicle.systemId,
          targetComponent: vehicle.componentId,
          defaultTimeout: commandTimeout,
        );

  final MavlinkSession session;
  final MavlinkNode vehicle;

  final ParameterProtocol parameters;
  final MissionProtocol mission;
  final CommandProtocol command;

  int get targetSystem => vehicle.systemId;
  int get targetComponent => vehicle.componentId;
}

/// Ground control station bootstrap: session, heartbeat publisher, and monitor.
///
/// Use [waitForVehicle] then [vehicleClient] to obtain protocol clients.
class MavlinkGcs {
  MavlinkGcs({
    required this.session,
    required this.heartbeatPublisher,
    required this.heartbeatMonitor,
  });

  final MavlinkSession session;
  final HeartbeatPublisher heartbeatPublisher;
  final HeartbeatMonitor heartbeatMonitor;

  /// Start heartbeat publish/monitor loops.
  void start() {
    heartbeatMonitor.start();
    heartbeatPublisher.start();
  }

  /// Stop heartbeat publish/monitor loops.
  Future<void> stopHeartbeats() async {
    heartbeatPublisher.stop();
    await heartbeatMonitor.stop();
  }

  /// Wait for the first vehicle and return a [MavlinkVehicleClient] for it.
  Future<MavlinkVehicleClient> waitForVehicle({
    Set<int>? excludeSystemIds,
    Duration timeout = const Duration(seconds: 60),
  }) {
    return heartbeatMonitor
        .waitForVehicle(excludeSystemIds: excludeSystemIds, timeout: timeout)
        .then((node) => MavlinkVehicleClient(session: session, vehicle: node));
  }

  /// Build a [MavlinkVehicleClient] for a known [vehicle] node.
  MavlinkVehicleClient vehicleClient(MavlinkNode vehicle) {
    return MavlinkVehicleClient(session: session, vehicle: vehicle);
  }

  /// Factory for a typical GCS setup over [link].
  factory MavlinkGcs.connect({
    required MavlinkDialect dialect,
    required MavlinkLink link,
    int systemId = 255,
    int componentId = 190,
    Duration heartbeatInterval = const Duration(seconds: 1),
    Duration heartbeatTimeout = const Duration(seconds: 3),
  }) {
    final session = MavlinkSession(
      dialect: dialect,
      link: link,
      systemId: systemId,
      componentId: componentId,
    );

    return MavlinkGcs(
      session: session,
      heartbeatPublisher: HeartbeatPublisher(
        session: session,
        heartbeat: HeartbeatTemplates.gcs(mavlinkVersion: dialect.version),
        interval: heartbeatInterval,
      ),
      heartbeatMonitor: HeartbeatMonitor(session: session, timeout: heartbeatTimeout),
    );
  }

  Future<void> close() async {
    await stopHeartbeats();
    await session.close();
  }
}
